PCDLoader with Worker support: parseAsync and run

kaisalmen opened the pull request:
This whole work was motivated by two things:

1.  I wanted to apply the whole common functionality (Builder, WorkerSupport, WorkerDirector, LoaderBase) I extracted in the past from `OBJLoader2` into `LoaderSupport` and show that it really works in a different loader.  `PCDLoader` was a fairly easy target.

2. Someone really asked for this feature (https://github.com/kaisalmen/WWOBJLoader/issues/10) months before `OBJLoader2` V2.x.x was completed and available here (R88), but I only lately had time to work on it.

What has been done to `PCDLoader`?
 - The Parser has been isolated into its own object/class. This can be used inside the worker.
 - The loader is based on `THREE.LoaderSupport.LoaderBase` to be able to re-use common functions (load function became completely generic)
 - `THREE.LoaderSupport.Builder` is commonly used to build meshes
 - `THREE.LoaderSupport.WorkerSupport` is used to build the Worker code and handle execution
 - It is now possible to use `THREE.LoaderSupport.WorkerDirector` to automate loading of files (my idea is to enhance the obj2_parallels example with two or more different loaders.

Two important things:
 - The code requires the latest `LoaderSupport` from #13156  that why this PR is based on that branch
 -  #13197 needs to be fixed otherwise the one of the three identical PCDs is not loaded 

Please review, let me know what you think?
Thanks you,

Kai

Mugen87 commented:
TBH, i don't vote for this change. The thing is... if you now start with `PCDLoader`, it won't take long until other loaders will be enhanced by `LoaderSupport`. And i don't think that is necessary. I prefer to keep the loaders simple. If users really need to use a worker for loading stuff, they can have a look at `OBJLoader2`.

looeee commented:
At first glance I agree with @Mugen87. In my opinion we would have to demonstrate that there will be a significant performance gain here before we consider adding this extra complexity. 

@kaisalmen I've taken a look through your previous PRs related to `WWOBJLoader` and I don't see any discussion about performance gains. Apologies if I missed it somewhere, have you benchmarked the difference `OBJLoader2` with and without workers?

kaisalmen commented:
@Mugen87 @looeee using a worker will not increase parsing speed compared to the same the Parser code running without the worker. It will be slower due to setup and data exchanges. This is true for `OBJLoader2` as well. 
What you gain is that the main thread is not blocked while parsing of large files is ongoing. That was the main motivation for `OBJLoader2`.

I can understand that you don't like the increase in code complexity. From the user of the Loader this is totally hidden as the method signatures are backward-compatible or identical.

From my pint of view a gain is the isolation of the actual Parser into its own object. A possible way forward could be: The `PCDLoader` stays almost as is, just the Parser is isolated and then I would add an independent new file which extends `PCDLoader` and use `LoaderSupport` or parts of it.

looeee commented:
> What you gain is that the main thread is not blocked while parsing of large files is ongoing.

Ah, I see. That's actually a really nice improvement. I've spent quite a bit of time trying to get loading overlays to run smoothly! 

What kind of performance decrease are we looking at though? Have you benchmarked it?

If it can be made entirely transparent to the user _and_ if we can avoid cluttering the loader with worker code then this could be a good addition. 

This PR which seems to do a of other stuff related to `OBJLoader`, docs and examples rather than the `PCDLoader`, could you separate those into a different PR so that it's clearer what's going on here?

kaisalmen commented:
@looeee performance decrease exists on small files due to init and worker/main communication overhead, but it becomes a minimal increase on large files (see below).  These are numbers obtained from `OBJLoader2` (Chrome Dev 65, Windows 10, Core i7-6700@3.4 Ghz, GTX 1080Ti): 

Overall Stats from console (1 MB file):
```console
Global output object count: 1
Overall counts: 
	Vertices: 48480
	Faces: 48480
```
Without worker (first is pure time spent inside Parser, second overall time):
```console
OBJLoader2.Parser.parse: 59.8759765625ms
OBJLoader2 parse: WaltHead: 62.263916015625ms
```
With worker (first is pure time spent inside Parser (worker), second overall time (main)):
```console
OBJLoader2.Parser.parse: 70.421142578125ms
OBJLoader2 parseAsync: WaltHead: 93.399169921875ms
```

On large files the Worker approach is minimally faster, but the gain is smaller than 5% and within nominal measurement error range.

Overall Stats from console (150 MB file):
```console
Global output object count: 206
Overall counts: 
	Vertices: 7905534
	Faces: 7905534
```
Without worker (first is pure time spent inside Parser, second overall time):
```console
OBJLoader2.Parser.parse: 7222.063232421875ms
OBJLoader2 parse: oven: 7224.68408203125ms
```
With worker (first is pure time spent inside Parser (worker), second overall time (main)):
```console
OBJLoader2.Parser.parse: 6989.02392578125ms
OBJLoader2 parseAsync: oven: 7092.661865234375ms
````

kaisalmen commented:
> This PR which seems to do a of other stuff related to OBJLoader, docs and examples rather than the PCDLoader, could you separate those into a different PR so that it's clearer what's going on here

@looeee The PR needs not yet merged changes from #13156 that why the list is so long. Only the last commit is relevant for `PCDLoader`.

> I prefer to keep the loaders simple.

@Mugen87 the question is why do you think it is not simple?
The only things really that are really imposed on a Loader by `LoaderSupport` are:
- The Parser must be isolated in its own object (I already have an idea how to remove this constraint...)
- The Parser must have a parse method that takes an ArrayBuffer
- A function must be invoked to build the mesh. This hook is required for the Parser running inside the worker to push buffers and other params back to the main thread. When it runs on the main, it just calls the `THREE.LoaderSupport.Builder` directly, but that must not be the case even. The parser logic itself is not really affected.

Additonally, I guess you think imposing a base prototype (`LoaderBase`) is too restrictive. Yes, that may have been over the top as it is really not required here.

Some thoughts: What locally does not look simple, appears simple if you widen the scope. If multiple loaders work by the outlined convention, then they all use common functions to gain worker capability. There is no need to write things n-times and fix it n-times.

I guess, there is need for discussion in a new issue asking if there really is the desire to use `WorkerSupport`, `Builder` and `WorkerDirector` elsewhere and if does not fit the needs how to adopt it, if desired. What exist today is already the outcome of multiple discussions including the convention outlined above, so let's see then...

Did I miss anything?

kaisalmen committed:
Alter PCDLoader to allow creation of extended worker based loader:
- Parser is isolated and mesh creation is confined in one method
- Usage of THREE features outside of buildMesh is limited to THREE.LoaderUtils

kaisalmen commented:
@Mugen87 are you ok with these less drastic changes to `PCDLoader` (see last forced commit)?
- Parser is isolated and mesh creation is confined in one method
- Usage of THREE features outside of buildMesh is limited to THREE.LoaderUtils

These changes will allow the creation of an extended loader (TODO for me) that then uses `LoaderSupport` adding worker support.

I kept the original proposal here: https://github.com/kaisalmen/three.js/tree/PCDLoader_ref

kaisalmen committed:
#13263 Introduce WorkerPCDLoader: It extends the existing loader and relies on LoaderSupport for embedding the Parser in a Worker.
LoaderSupport needed small enhancements allowing to construct extended prototypes.
TODO:
- Dedicated example and revert of current
- Introduce run method in WorkerPCDLoader

kaisalmen commented:
@Mugen87 and @looeee I have created `WorkerPCDLoader` that extends the slightly altered `PCDLoader`. Only the extended loader requires `LoaderSupport`. It takes the isolated Parser from `PCDLoader` overrides the `buildMesh` functions and packages everything into a Blob that becomes the Worker.
A new example is still missing...

looeee commented:
This is a step in the right direction I think. However in my opinion `LoaderSupport` still does too much. Is all the console logging / validator stuff required for worker support? 

I would rather something like `WorkerLoader.js` that converts a suitable `Loader` into a `WorkerLoader`. The API could look something like:

```js
const PCDWorkerLoader = new THREE.WorkerLoader( new THREE.PCDLoader() );

const OBJWorkerLoader = new THREE.WorkerLoader( new THREE.OBJLoader() );
```

This way we would avoid cluttering the examples with multiple versions of each loader. 
Ideally this could be made to work with more complex formats like GLTF / FBX / Collada that support animation / skinning / morph targets and so on. 

Is this possible? I'm not hugely familiar with using web workers so let me know if I'm suggesting something that's unreasonable 😄

donmccurdy commented:
> Ideally this could be made to work with more complex formats like GLTF / FBX / Collada that support animation / skinning / morph targets and so on.

FWIW, my gut feeling is that running GLTFLoader in a Web Worker won't make much sense: it's already just parsing JSON and creating views into ArrayBuffers (unlike e.g. ColladaLoader which must do some heavy lifting with XML) so parsing whatever Transferrable the Web Worker creates would likely require similar time. But, another example that I'd _love_ to get into a web worker is `THREE.DRACOLoader`... GLTFLoader uses that internally, so replacing `gltfLoader.setDracoLoader( loader )` with `gltfLoader.setDracoLoader( wwLoader )` as a drop-in substitute would be amazing. 🙂

kaisalmen commented:
@looeee I like your idea and actually `WorkerDirector` does something generic already. It creates an object (the Loader) from a given name and passes on a preloaded `WorkerSupport` (Worker Blob could therefore be re-used, init once, run n-times) to the `run` function for instruction based automation. The Loader can be anything as long as supports this contract. If we simplify this contract to `parseAsync` is required then it could become something like this:
```javascript
var loaderPcd = new THREE.WorkerLoader( THREE.PCDLoader );
var loaderObj = new THREE.WorkerLoader( THREE.OBJLoader );

// Director would then in addition require a function like `run` for passing automation instructions
var director = new THREE.WorkerDirector( [ loaderPcd, loaderObj ] );
```
`OBJLoader2` without a dependency to `LoaderSupport` will be anyway required for replacing `OBJLoader´ in the future, I guess.

> This is a step in the right direction I think. However in my opinion LoaderSupport still does too much. Is all the console logging / validator stuff required for worker support?
Not really, `WorkerSupport` and `WorkerRunnerRefImpl` could be easily made independent of `ConsoleLogger` and `Validator`. The later change will just make the code more verbose.

`LoaderSupport` is an umbrella of useful and maybe not so useful things. It should be split apart and beforehand inter-dependencies of functions should be cleared. I wanted to prevent loading of multiple files when `OBJLoader2` was introduced.
Still, the `Validator` is very useful checking legimate content on **public** functions. I wonder whether such util functions shouldn't be part of the core library.
Regarding `ConsoleLogger`, honestly, the only really good thing is that debug and info level logging can be completely disabled. Is there a generic approach to this in three.js? I would immediately replace this with something more suited.

For anyone interested, this is how the generated worker Blob of `WorkerPCDLoader` looks like (already now `WorkerSupport` allows to replace `WorkerRunnerRefImpl`with your own work communication contract):
<details>
/**
  * This code was constructed by WorkerPCDLoader.buildCode.
  */

THREE = {
	LoaderSupport: {},
	PCDLoader: {},
	WorkerPCDLoader: {}
};

THREE.LoaderUtils = {
	decodeText: function ( array ) {

			if ( typeof TextDecoder !== 'undefined' ) {

				return new TextDecoder().decode( array );

			}

			// Avoid the String.fromCharCode.apply(null, array) shortcut, which
			// throws a "maximum call stack size exceeded" error for large arrays.

			var s = '';

			for ( var i = 0, il = array.length; i < il; i ++ ) {

				// Implicitly assumes little-endian.
				s += String.fromCharCode( array[ i ] );

			}

			// Merges multi-byte utf-8 characters.
			return decodeURIComponent( escape( s ) );

		},
	extractUrlBase: function ( url ) {

			var parts = url.split( '/' );

			if ( parts.length === 1 ) return './';

			parts.pop();

			return parts.join( '/' ) + '/';

		},
}

THREE.PCDLoader.Parser = (function () {

	function Parser () {
	this.littleEndian = true;
};

	Parser.prototype.setLittleEndian = function ( littleEndian ) {
		this.littleEndian = littleEndian === true;
	};

	Parser.prototype.parse = function ( data ) {
		var textData = THREE.LoaderUtils.decodeText( data );
		var pcdHeader = this.parseHeader( textData );
		return this.parseData( pcdHeader, textData, data );
	};

	Parser.prototype.parseHeader = function ( data ) {

		var PCDheader = {};
		var result1 = data.search( /[\r\n]DATA\s(\S*)\s/i );
		var result2 = /[\r\n]DATA\s(\S*)\s/i.exec( data.substr( result1 - 1 ) );

			PCDheader.data = result2[ 1 ];
			PCDheader.headerLen = result2[ 0 ].length + result1;
			PCDheader.str = data.substr( 0, PCDheader.headerLen );

		// remove comments

			PCDheader.str = PCDheader.str.replace( /\#.*/gi, '' );

		// parse

			PCDheader.version = /VERSION (.*)/i.exec( PCDheader.str );
			PCDheader.fields = /FIELDS (.*)/i.exec( PCDheader.str );
			PCDheader.size = /SIZE (.*)/i.exec( PCDheader.str );
			PCDheader.type = /TYPE (.*)/i.exec( PCDheader.str );
			PCDheader.count = /COUNT (.*)/i.exec( PCDheader.str );
			PCDheader.width = /WIDTH (.*)/i.exec( PCDheader.str );
			PCDheader.height = /HEIGHT (.*)/i.exec( PCDheader.str );
			PCDheader.viewpoint = /VIEWPOINT (.*)/i.exec( PCDheader.str );
			PCDheader.points = /POINTS (.*)/i.exec( PCDheader.str );

		// evaluate

			if ( PCDheader.version !== null )
				PCDheader.version = parseFloat( PCDheader.version[ 1 ] );

			if ( PCDheader.fields !== null )
				PCDheader.fields = PCDheader.fields[ 1 ].split( ' ' );

			if ( PCDheader.type !== null )
				PCDheader.type = PCDheader.type[ 1 ].split( ' ' );

			if ( PCDheader.width !== null )
				PCDheader.width = parseInt( PCDheader.width[ 1 ] );

			if ( PCDheader.height !== null )
				PCDheader.height = parseInt( PCDheader.height[ 1 ] );

			if ( PCDheader.viewpoint !== null )
				PCDheader.viewpoint = PCDheader.viewpoint[ 1 ];

			if ( PCDheader.points !== null )
				PCDheader.points = parseInt( PCDheader.points[ 1 ], 10 );

			if ( PCDheader.points === null )
				PCDheader.points = PCDheader.width * PCDheader.height;

			if ( PCDheader.size !== null ) {

				PCDheader.size = PCDheader.size[ 1 ].split( ' ' ).map( function ( x ) {

				return parseInt( x, 10 );

			} );

		}

			if ( PCDheader.count !== null ) {

				PCDheader.count = PCDheader.count[ 1 ].split( ' ' ).map( function ( x ) {

				return parseInt( x, 10 );

			} );

		} else {

				PCDheader.count = [];

				for ( var i = 0, l = PCDheader.fields.length; i < l; i ++ ) {

					PCDheader.count.push( 1 );

			}

		}

			PCDheader.offset = {};

		var sizeSum = 0;

			for ( var i = 0, l = PCDheader.fields.length; i < l; i ++ ) {

				if ( PCDheader.data === 'ascii' ) {

					PCDheader.offset[ PCDheader.fields[ i ] ] = i;

			} else {

					PCDheader.offset[ PCDheader.fields[ i ] ] = sizeSum;
					sizeSum += PCDheader.size[ i ];

			}

		}

		// for binary only
		PCDheader.rowSize = sizeSum;

		return PCDheader;
	};

	Parser.prototype.parseData = function ( PCDheader, textData, data ) {
		var position = [];
		var normal = [];
		var color = [];

		// ascii

		if ( PCDheader.data === 'ascii' ) {

			var offset = PCDheader.offset;
			var pcdData = textData.substr( PCDheader.headerLen );
			var lines = pcdData.split( '\n' );

			for ( var i = 0, l = lines.length; i < l; i ++ ) {

				var line = lines[ i ].split( ' ' );

				if ( offset.x !== undefined ) {

					position.push( parseFloat( line[ offset.x ] ) );
					position.push( parseFloat( line[ offset.y ] ) );
					position.push( parseFloat( line[ offset.z ] ) );

				}

				if ( offset.rgb !== undefined ) {

					var c = new Float32Array( [ parseFloat( line[ offset.rgb ] ) ] );
					var dataview = new DataView( c.buffer, 0 );
					color.push( dataview.getUint8( 0 ) / 255.0 );
					color.push( dataview.getUint8( 1 ) / 255.0 );
					color.push( dataview.getUint8( 2 ) / 255.0 );

				}

				if ( offset.normal_x !== undefined ) {

					normal.push( parseFloat( line[ offset.normal_x ] ) );
					normal.push( parseFloat( line[ offset.normal_y ] ) );
					normal.push( parseFloat( line[ offset.normal_z ] ) );

				}

			}

		}

		// binary

		if ( PCDheader.data === 'binary_compressed' ) {

			console.error( 'THREE.PCDLoader: binary_compressed files are not supported' );
			return;

		}

		if ( PCDheader.data === 'binary' ) {

			var dataview = new DataView( data, PCDheader.headerLen );
			var offset = PCDheader.offset;

			for ( var i = 0, row = 0; i < PCDheader.points; i ++, row += PCDheader.rowSize ) {

				if ( offset.x !== undefined ) {

					position.push( dataview.getFloat32( row + offset.x, this.littleEndian ) );
					position.push( dataview.getFloat32( row + offset.y, this.littleEndian ) );
					position.push( dataview.getFloat32( row + offset.z, this.littleEndian ) );

				}

				if ( offset.rgb !== undefined ) {

					color.push( dataview.getUint8( row + offset.rgb + 0 ) / 255.0 );
					color.push( dataview.getUint8( row + offset.rgb + 1 ) / 255.0 );
					color.push( dataview.getUint8( row + offset.rgb + 2 ) / 255.0 );

				}

				if ( offset.normal_x !== undefined ) {

					normal.push( dataview.getFloat32( row + offset.normal_x, this.littleEndian ) );
					normal.push( dataview.getFloat32( row + offset.normal_y, this.littleEndian ) );
					normal.push( dataview.getFloat32( row + offset.normal_z, this.littleEndian ) );

				}

			}

		}

		return this.buildMesh( position, normal, color );
	};

	return Parser;
})();

THREE.WorkerPCDLoader.Parser = (function () {


Parser.prototype = Object.create( THREE.PCDLoader.Parser.prototype );
Parser.constructor = Parser;

	function Parser ( manager ) {

	THREE.PCDLoader.Parser.call( this );
	this.callbackBuilder = null;

};

	Parser.prototype.setCallbackBuilder = function ( callbackBuilder ) {
	if ( callbackBuilder === null || callbackBuilder === undefined ) throw 'Unable to run as no "builder" callback is set.';
	this.callbackBuilder = callbackBuilder;
};

	Parser.prototype.buildMesh = function ( position, normal, color ) {
	var vertexFA = new Float32Array( position );
	var normalFA = normal.length > 0 ? new Float32Array( normal ) : null;
	var colorFA = color.length > 0 ? new Float32Array( color ) : null;

	var mesh = this.callbackBuilder( {
			cmd: 'meshData',
			progress: {
				numericalValue: 100
			},
			params: {},
			materials: {
				multiMaterial: false,
				materialNames: [ color.length > 0 ? 'defaultVertexColorMaterial' : 'defaultPointMaterial' ],
				materialGroups: null
			},
			buffers: {
				vertices: vertexFA,
				indices: null,
				colors: colorFA,
				normals: normalFA,
				uvs: null
			},
			// 0: mesh, 1: line, 2: point
			geometryType: 2
		},
		[ vertexFA.buffer ],
		null,
		colorFA !== null ? [ colorFA.buffer ] : null,
		normalFA !== null ? [ normalFA.buffer ] : null,
		null
	);

	return mesh;
};

	Parser.prototype.parse = function ( data ) {
		var textData = THREE.LoaderUtils.decodeText( data );
		var pcdHeader = this.parseHeader( textData );
		return this.parseData( pcdHeader, textData, data );
	};

	return Parser;
})();

var Parser = THREE.WorkerPCDLoader.Parser;

WorkerRunnerRefImpl = (function () {

	function WorkerRunnerRefImpl() {
		var scope = this;
		var scopedRunner = function( event ) {
			scope.processMessage( event.data );
		};
		self.addEventListener( 'message', scopedRunner, false );
	}

	WorkerRunnerRefImpl.prototype.applyProperties = function ( parser, params ) {
		var property, funcName, values;
		for ( property in params ) {
			funcName = 'set' + property.substring( 0, 1 ).toLocaleUpperCase() + property.substring( 1 );
			values = params[ property ];

			if ( typeof parser[ funcName ] === 'function' ) {

				parser[ funcName ]( values );

			} else if ( parser.hasOwnProperty( property ) ) {

				parser[ property ] = values;

			}
		}
	};

	WorkerRunnerRefImpl.prototype.processMessage = function ( payload ) {
		var logEnabled = payload.logger.enabled;
		var logDebug = payload.logger.enabled;
		if ( payload.cmd === 'run' ) {

			var callbacks = {
				callbackBuilder: function ( payload ) {
					self.postMessage( payload );
				},
				callbackProgress: function ( text ) {
					if ( logEnabled && logDebug ) console.debug( 'WorkerRunner: progress: ' + text );
				}
			};

			// Parser is expected to be named as such
			var parser = new Parser();
			if ( typeof parser[ 'setLogConfig' ] === 'function' ) parser.setLogConfig( logEnabled, logDebug );
			this.applyProperties( parser, payload.params );
			this.applyProperties( parser, payload.materials );
			this.applyProperties( parser, callbacks );
			parser.workerScope = self;
			parser.parse( payload.data.input, payload.data.options );

			if ( logEnabled ) console.log( 'WorkerRunner: Run complete!' );

			callbacks.callbackBuilder( {
				cmd: 'complete',
				msg: 'WorkerRunner completed run.'
			} );

		} else {

			console.error( 'WorkerRunner: Received unknown command: ' + payload.cmd );

		}
	};

	return WorkerRunnerRefImpl;
})();

new WorkerRunnerRefImpl();
</details>

kaisalmen commented:
@donmccurdy concerning `DracoLoader` I see at least the following possibilities:
- Isolate the Decoder insides the Loader and make it independent of THREE dependencies (e.g. `BufferGeometry`), exchange buffers/transferables and create meshes in outside worker (`OBJLoader2` approach with `LoaderSupport.Builder` in Main). **Verdict:** Likely most work, but short worker creation time, smallest possible worker code.
- Leave code as, include THREE in worker and after decoding is done disassemble meshes to buffers, transfer to Main and create meshes here again. **Verdict:** Only extra code needed is the worker mesh disassemble wrapper, long worker creation time and big worker size. Some work is done twice, but decoding done in parallel to Main. Becomes efficient if re-used.

Don't know if the analysis is correct for the technology chain (wasm), but if it is compatible with worker execution, then it should not invalidate the outlined approaches.

Maybe, there is really the need to discuss the why, how, what and where concerning workers in a separate issue...

kaisalmen commented:
@looeee `ConsoleLogger` and `LoaderBase` will be gone in R91. PR was merged a couple of hours ago. I will come up with a proposal for a generic `WorkerLoader` for R92.

looeee commented:
Sweet 😁

kaisalmen commented:
@Mugen87 I will close this. It will be available for free once #13664 is available.
